### ISSUE TITLE
memoir class crashes LaTeX

### ISSUE THREAD

**Issue by jfbu** [sphinx-doc/sphinx#2629]
Using 

``` python
latex_elements = {
    'docclass': 'memoir',
}
latex_documents = [
    (master_doc, 'testmemoir.tex', 'testmemoir Documentation',
     'jfbu', 'manual', True),
]
```

in conf.py leads to correct header of latex source:

``` latex
% Generated by Sphinx.
\def\sphinxdocclass{memoir}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
```

but then compilation fails:

``` latex
! LaTeX Error: Command \titleref already defined.
               Or name \end... illegal, see p.192 of the manual.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              

l.152 \newcommand{\titleref}[1]{\emph{#1}}
```

There is some special support code in  sphinx.sty, sphinxmanual.cls and sphinxhowto.cls for memoir class. But currently it does not work. Should Sphinx really support memoir ?

**Comment by tk0miya** [comment]
Oh, it's my mistake for naming. I used very popular words to define the macro.
It causes the conflict. It's not my intension.
So we have to rename it ane make it unique.

Is there good naming custom or techniques for LaTeX packages?

**Comment by jfbu** [comment]
@tk0miya there is no namespace with TeX/LaTeX macros (in the future LaTeX3 will perhaps improve a bit the situation, making it mandatory for package authors to pick up a prefix for all macros).

It was typical of twenty years ago that people were using common words such as `\formula`, `\bold`, `\num`, `\link`, not only for user interface but also internally. Sometimes it is successful like `\foreach` from TikZ, because it wins by Darwinian selection, but sometimes using common word is only a nuisance to others.

Over the years, better habits have developed and most packages now use a prefix at least for all their private macros. Typically the prefix is two or three letters, for example `\Hy@` for hyperref, `\LT@` for longtable (some things do not have the `@`), `\Gin@` for graphicx,  surprisingly there isn't a central database of "registered" prefix to avoid clash. The best is when the package uses a prefix for inner macros (typically with a `@` or more) and also one for public macros, but the latter is rarer. Sometimes people writing a package about footnotes use `\FN@` but this is bad because that's exactly the choice others would do too !

Currently some parts of `sphinx.sty` use `\py@` prefix, the pygments one use `\PYG..`,  more recent code tend to use `\Sphinx@` or `\Sphinx` or `\sphinx` or `\sphinx@` (I am a bit guilty there).

We could decide to use systematically `\SPH@` or `\sph@` prefix for inner macros, and `\Sphinx..` for public ones or for those the LaTeX writer needs to put inside the document, where the @ is not a letter. For example `\code` should better be something like `\sphinxcode` or `\SphinxCode` depending on taste. Perhaps  `\spxcode` for brevity, or `\sphxcode`. And `\titleref` could then be `\sphxtitleref` or  `\SphinxTitleRef`.

Also, color defintions etc... should have some Sphinx prefix too I guess.

**Comment by jfbu** [comment]
I have another question: the sphinx doc for latex options has

```
New in version 1.2: In the past including your own document class required you to prepend
the  document class name with the string “sphinx”. This is not necessary anymore.
```

and the quickstart conf.py has

```
# Grouping the document tree into LaTeX files. List of tuples
# (source start file, target name, title,
#  author, documentclass [howto, manual, or own class]).
```

I thus naively tried 'memoir' as fifth element. But in fact it should be done with `'docclass'`. On the other hand `'docclass'` is said later in the doc to be among

```
Keys that are set by other options and therefore should not be overridden
```

is it still correct that `'docclass'` key is set by an other option ? I do not quite understand the Python syntax of latex.py.

**Comment by jfbu** [comment]
Two things are needed for the user wishing to use memoir class:
- use the PR #2675 and add <strike>`latex_macros_onlywithprefix = True`</strike> `latex_keep_old_macro_names=False` in conf.py
- use the `howto` class, not `manual`. Indeed `sphinxmanual.cls` uses environment `titlepage` in `\maketitle` and this causes another clash with memoir. Here is extract from its doc:
  
  > This class provides neither a titlepage option nor a titlepage
  >  environment; instead it provides the titlingpage environment which falls
  >  between the titlepage option and the titlepage environment.

`sphinxmanual.cls` mentions memoir but is anyhow incompatible due to its use of `titlepage` environment. Only `sphinxhowto.cls` is compatible and can load `memoir.cls`.

**Comment by jfbu** [comment]
Fix by #2675, at 5510653 on stable branch (future 1.4.5).

**Comment by tk0miya** [comment]
Sorry. I've missed your question.

The framework for docclass is very complicated.

The 5th element of `latex_document` is used for an argument of `\documentclass` definition.
As a special case, the value equals to `howto` or `manual`, then `sphinxhowto` or `sphinxmanual` is used for the definition.

BTW, what the `docclass` key of `latex_elements` is used for? The answer is the value of `\sphinxdocclass`. It is used in `sphinxhowto` and `sphinxmanual` class.
To determine the value, sphinx uses `latex_docclass` and the 5th element of `latex_document`.

For example, the 5th elements of `latex_documents` is...
- case of `manual`: `\documentclass{sphinxmanual}` and `\def\sphinxdocclass{report}`
- case of `howto`: `\documentclass{sphinxhowto}` and `\def\sphinxdocclass{article}`
- case of `memoir`: `\documentlcass{memoir}` and `\def\sphinxdocclass{report}` (but never used)

Does this help your understanding?

**Comment by jfbu** [comment]
Yes, thanks. In this use of `memoir`, the `sphinx.sty` is not used, hence nothing can compile because the produced mark-up (like `notice` environment, `\strong` or `code` text styling) is not understood by memoir. This is why I used ``'docclass': 'memoir',`` in `latex_elements` and
```
latex_documents = [
    (master_doc, 'testmemoir.tex', 'testmemoir Documentation',
     'jfbu', 'howto'),
]
```
then latex has 
```latex
\def\sphinxdocclass{memoir}
\documentclass[letterpaper,10pt,english]{sphinxhowto}
```
and it works (after the fix). It wouldn't work with `sphinxmanual` due to `titlepage` issue.

**Comment by jfbu** [comment]
Yes, thanks. Sorry about a comment I have suppressed now, which was wrong. My use case was with 
`'docclass': 'memoir',` in `latex_elements` and 

```
latex_documents = [
    (master_doc, 'testmemoir.tex', 'testmemoir Documentation',
     'jfbu', 'howto'),
]
```

then latex file starts with 

``` latex
\def\sphinxdocclass{memoir}
\documentclass[letterpaper,10pt,english]{sphinxhowto}
```

and the `sphinxhowto.cls` loads `memoir` class, as this seems how it should be. Is not possible with `manual` because of `titlepage` issue. In your example with `memoir`, it is `memoir` class which loads  `sphinx.sty`. Somehow I didn't try that (perhaps because I got an error initially).
